File: java/leetcode/generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList();
        backtrack(ans, "", 0, 0, n);
        return ans;
    }
    
    public void backtrack(List<String> ans, String current, int open, int close, int max) {
        if (current.length() == max * 2) {
            ans.add(current);
            return;
        }

        if (open < max) {
            backtrack(ans, current + '(', open + 1, close, max);
        }
        if (close < open) {
            backtrack(ans, current + ')', open, close + 1, max);
        }
    }
}

File: javascript/leetcode/generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    const ans = [];
    var backtrack = function(current, open, close, max) {
        if (current.length === max * 2) {
            ans.push(current);
            return
        }

        if (open < max) {
            backtrack(current + '(', open + 1, close, max);
        }

        if (close < open) {
            backtrack(current + ')', open, close + 1, max);
        }
    }
    backtrack('', 0, 0, n);
    return ans;
};
